--- SConscript	2009-03-25 11:08:54.282021710 -0400
+++ SConscript.new	2009-03-25 11:08:48.227021626 -0400
@@ -600,7 +600,7 @@
                     Action(create_h_wrapper),
                     Delete('${TARGETS[1]}pp')])
 
-BuildYacc(env, 'xml', 'xpathyy', 'XPathGrammar')
+BuildYacc(env, 'xml2', 'xpathyy', 'XPathGrammar')
 BuildYacc(env, 'css', 'cssyy',   'CSSGrammar')
 
 # TODO(bradnelson): need to add in error checking
