--- SConscript.main	2009-03-29 23:29:22.197377684 -0400
+++ SConscript.main.new	2009-03-30 13:01:31.248404386 -0400
@@ -594,7 +594,8 @@
 # Build with system-provided NSS and GTK.
 if root_env['PLATFORM'] in ['linux', 'linux2', 'posix']:
   try:
-    linux_env.ParseConfig('pkg-config --cflags --libs nss')
+    linux_env.Append(CCFLAGS=['-I/usr/include/nss', '-I/usr/include/nspr', '-Wl,-R/usr/lib32/nss', '-Wl,-R/usr/lib32/nspr'])
+    linux_env.Append(LINKFLAGS=['-L/usr/lib32/nss', '-L/usr/lib32/nspr', '-lssl3', '-lsmime3', '-lnssutil3', '-lnss3', '-lplds4', '-lplc4', '-lnspr4', '-lpthread', '-ldl'])
     linux_env.ParseConfig('pkg-config --cflags --libs gtk+-2.0')
     linux_env.ParseConfig('pkg-config --cflags --libs pangoft2')
   except OSError, e:
