diff --git a/SConscript b/SConscript
index 1e7c797..48cd568 100755
--- a/SConscript
+++ b/SConscript
@@ -128,7 +128,7 @@ env_dll.Append(
         'renderer',
         'hunspell',
         'sdch',
-        'sqlite',
+        'sqlite3',
         'v8_snapshot',
         'views',
 
diff --git a/common/ipc_tests.scons b/common/ipc_tests.scons
index 9625eab..5381134 100755
--- a/common/ipc_tests.scons
+++ b/common/ipc_tests.scons
@@ -46,7 +46,7 @@ if env.Bit('posix'):
         'browser',
         'port',
         'renderer',
-        'sqlite',
+        'sqlite3',
       ]
   )
 
diff --git a/test/interactive_ui/interactive_ui_tests.scons b/test/interactive_ui/interactive_ui_tests.scons
index dfe56a9..de37127 100755
--- a/test/interactive_ui/interactive_ui_tests.scons
+++ b/test/interactive_ui/interactive_ui_tests.scons
@@ -51,7 +51,7 @@ env.Prepend(
         'plugin',
         'port',
         'renderer',
-        'sqlite',
+        'sqlite3',
         'test_chrome_plugin',
         'util',
         'views',
diff --git a/test/page_cycler/page_cycler_tests.scons b/test/page_cycler/page_cycler_tests.scons
index be56817..db63c77 100755
--- a/test/page_cycler/page_cycler_tests.scons
+++ b/test/page_cycler/page_cycler_tests.scons
@@ -34,7 +34,7 @@ env.Prepend(
         'glue',
         'WTF',
         # needed for shared link deps
-        'sqlite',
+        'sqlite3',
         'xslt',
     ],
 )
diff --git a/test/perf/perftests.scons b/test/perf/perftests.scons
index 031f386..a2ad5cc 100755
--- a/test/perf/perftests.scons
+++ b/test/perf/perftests.scons
@@ -45,7 +45,7 @@ env.Prepend(
 
         'browser',
         'common',
-        'sqlite',
+        'sqlite3',
 
         'JavaScriptCore_pcre',
         'V8Bindings',
diff --git a/test/startup/startup_tests.scons b/test/startup/startup_tests.scons
index 7150bca..11594f1 100755
--- a/test/startup/startup_tests.scons
+++ b/test/startup/startup_tests.scons
@@ -34,7 +34,7 @@ env.Prepend(
         'glue',
         'WTF',
         # needed for shared link deps
-        'sqlite',
+        'sqlite3',
         'xslt',
     ],
 )
diff --git a/test/ui/ui_tests.scons b/test/ui/ui_tests.scons
index 2736916..d47bf9b 100644
--- a/test/ui/ui_tests.scons
+++ b/test/ui/ui_tests.scons
@@ -40,7 +40,7 @@ env.Prepend(
         'common',
         'glue',
         'port',
-        'sqlite',
+        'sqlite3',
         'WTF',
         'views',
     ]
diff --git a/test/unit/unit_tests.scons b/test/unit/unit_tests.scons
index edc4ed7..4bcfd6d 100644
--- a/test/unit/unit_tests.scons
+++ b/test/unit/unit_tests.scons
@@ -45,7 +45,7 @@ env.Prepend(
         'debugger',
         'plugin',
         'renderer',
-        'sqlite',
+        'sqlite3',
         'util',
         'v8_snapshot',
 
diff --git a/tools/profiles/generate_profile.scons b/tools/profiles/generate_profile.scons
index cead552..928c248 100755
--- a/tools/profiles/generate_profile.scons
+++ b/tools/profiles/generate_profile.scons
@@ -42,7 +42,7 @@ env.Prepend(
         'plugin',
         'port',
         'renderer',
-        'sqlite',
+        'sqlite3',
         'util',
         'v8_snapshot',
         'V8Bindings',
