diff -Nurp gnome-do-0.6.1.0/Do/src/Do.cs gnome-do-0.6.1.0.new/Do/src/Do.cs
--- gnome-do-0.6.1.0/Do/src/Do.cs	2008-09-23 16:29:33.000000000 +0800
+++ gnome-do-0.6.1.0.new/Do/src/Do.cs	2008-12-14 10:46:18.000000000 +0800
@@ -89,8 +89,9 @@ namespace Do {
 				Controller.Summon ();
 			
 			AppDomain.CurrentDomain.ProcessExit += OnProcessExit;
-				
-			Gtk.Application.Run ();
+			try {	
+				Gtk.Application.Run ();
+			} catch (Exception e) { }
 		}
 		
 		static void OnProcessExit (object o, EventArgs args)
