Index: trunk/nm-applet.conf
===================================================================
--- trunk.orig/nm-applet.conf
+++ trunk/nm-applet.conf
@@ -11,7 +11,7 @@
 		<!-- Only root can get secrets -->
 		<allow send_interface="org.freedesktop.NetworkManagerSettings.Secrets"/>
 	</policy>
-	<policy at_console="true">
+	<policy group="plugdev">
 		<allow own="org.freedesktop.NetworkManagerUserSettings"/>
 
 		<allow send_destination="org.freedesktop.NetworkManagerUserSettings"/>
