--- gnome-mplayer-0.9.3.orig/Makefile.am	2008-09-30 08:07:53.000000000 -0600
+++ gnome-mplayer-0.9.3/Makefile.am	2009-01-06 22:41:17.000000000 -0700
@@ -38,7 +38,6 @@
 
 install-schemas: $(schema_DATA)
 	GCONF_CONFIG_SOURCE=$(GCONF_SCHEMA_CONFIG_SOURCE) gconftool-2 --makefile-install-rule $(schema_DATA) 2>&1 > /dev/null
-	gconftool-2 --shutdown
 
 if GCONF_SCHEMAS_INSTALL
 install-data-hook: install-schemas
