diff -Naur pyclutter-0.7.99_pre3288.orig/docs/Makefile.am pyclutter-0.7.99_pre3288/docs/Makefile.am
--- pyclutter-0.7.99_pre3288.orig/docs/Makefile.am	2008-08-04 18:03:21.000000000 +0530
+++ pyclutter-0.7.99_pre3288/docs/Makefile.am	2008-10-02 15:53:26.000000000 +0530
@@ -82,7 +82,7 @@
 	fi
 
 install-data-hook:
-	@$(PYGOBJECT_FIXXREF) -i $(PYGOBJECT_PYGDOCS) $(DESTDIR)$(TARGET_DIR)
+	python $(PYGOBJECT_FIXXREF) -i $(PYGOBJECT_PYGDOCS) $(DESTDIR)$(TARGET_DIR)
 
 uninstall-local:
 	rm -rf $(DESTDIR)$(TARGET_DIR)/*
