diff -Nrp genkernel-3.4.10.903/arch/x86/modules_load genkernel-3.4.10.903.new/arch/x86/modules_load
*** genkernel-3.4.10.903/arch/x86/modules_load	2009-01-06 06:22:10.000000000 +0800
--- genkernel-3.4.10.903.new/arch/x86/modules_load	2009-01-07 19:42:11.000000000 +0800
*************** MODULES_USB="ehci-hcd uhci usb-ohci hid 
*** 24,26 ****
--- 24,27 ----
  # Filesystems
  MODULES_FS="ext2 ext3 reiserfs jfs nfs xfs fuse"
  
+ MODULES_VIDEO="cfbcopyarea cfbfillrect cfbimgblt uvesafb"
diff -Nrp genkernel-3.4.10.903/defaults/linuxrc genkernel-3.4.10.903.new/defaults/linuxrc
*** genkernel-3.4.10.903/defaults/linuxrc	2009-01-06 06:22:10.000000000 +0800
--- genkernel-3.4.10.903.new/defaults/linuxrc	2009-01-07 19:42:56.000000000 +0800
***************
*** 3,8 ****
--- 3,11 ----
  . /etc/initrd.defaults
  . /etc/initrd.scripts
  
+ #915
+ 915resolution 58 1280 800 32
+ 
  splash() {
  	return 0
  }
*************** then
*** 220,225 ****
--- 223,237 ----
  	REAL_ROOT="${FAKE_ROOT}"	
  fi
  
+ real_mod_path() {
+ 	find /lib/modules | grep /"${1}${KSUFF}"
+ }
+ 
+ ${INSMOD} `real_mod_path cfbcopyarea`
+ ${INSMOD} `real_mod_path cfbfillrect`
+ ${INSMOD} `real_mod_path cfbimgblt`
+ ${INSMOD} `real_mod_path uvesafb` mode_option=1280x800-32 mtrr=3 scroll=ywrap
+ 
  splash 'init'
  
  detect_sbp2_devices
diff -Nrp genkernel-3.4.10.903/gen_initramfs.sh genkernel-3.4.10.903.new/gen_initramfs.sh
*** genkernel-3.4.10.903/gen_initramfs.sh	2009-01-06 06:22:10.000000000 +0800
--- genkernel-3.4.10.903.new/gen_initramfs.sh	2009-01-07 19:43:45.000000000 +0800
*************** append_busybox() {
*** 48,53 ****
--- 48,54 ----
  		gen_die 'Could not extract busybox bincache!'
  	chmod +x "${TEMP}/initramfs-busybox-temp/bin/busybox"
  
+ 	cp "/boot/backup/915resolution" ${TEMP}/initramfs-busybox-temp/bin/
  	mkdir -p "${TEMP}/initramfs-busybox-temp/usr/share/udhcpc/"
  	cp "${GK_SHARE}/defaults/udhcpc.scripts" ${TEMP}/initramfs-busybox-temp/usr/share/udhcpc/default.script
  	chmod +x "${TEMP}/initramfs-busybox-temp/usr/share/udhcpc/default.script"
