diff -Nurp gnome-do-plugins-0.6.0.1/Twitter/src/TwitterAction.cs gnome-do-plugins-0.6.0.1.new/Twitter/src/TwitterAction.cs
--- gnome-do-plugins-0.6.0.1/Twitter/src/TwitterAction.cs	2008-08-24 16:41:00.000000000 +0800
+++ gnome-do-plugins-0.6.0.1.new/Twitter/src/TwitterAction.cs	2008-12-14 10:29:59.000000000 +0800
@@ -56,20 +56,32 @@ namespace DoTwitter
 		
 		public static bool Connect (string username, string password)
 		{
+			System.Console.WriteLine("Tweet connect start");
 			twitter = new Twitter (username, password);
+			System.Console.WriteLine("Tweet connect end");
 			return true;
 		}
 		
 		public static List<IItem> Friends {
-			get { return items; }
+			get {
+				System.Console.WriteLine("Tweet get start");
+				return items; 
+				System.Console.WriteLine("Tweet get end");
+			}
 		}
 		
 		public static string Status {
-			set { status = value; }
+			set { 
+				System.Console.WriteLine("Tweet set start");
+				status = value; 
+				System.Console.WriteLine("Tweet set end");
+			}
 		}
 		
 		public static void Tweet ()
 		{
+			System.Console.WriteLine("Tweet start");
+
 			if (!(String.IsNullOrEmpty (username) || String.IsNullOrEmpty (password))) {
 				try {
 					twitter.Update (status);
@@ -89,10 +101,14 @@ namespace DoTwitter
 					"Tweet Failed"), Catalog.GetString ("Missing Login Credentials. "
 					+ "Please set login information in plugin configuration."));
 			}
+			System.Console.WriteLine("Tweet end");
+
 		}
 		
 		public static void UpdateFriends ()
 		{
+			System.Console.WriteLine("UpdateFriends start");
+
 			if ((String.IsNullOrEmpty (username) || String.IsNullOrEmpty (password)))
 				return;
 			
@@ -126,10 +142,12 @@ namespace DoTwitter
 			} finally {
 				Monitor.Exit (friends_lock);
 			} 
+			System.Console.WriteLine("UpdateFriends end");
 		}
 		
 		public static void UpdateTweets ()
 		{
+			System.Console.WriteLine("UpdateTweets, start");
 			if (!GenConfig.ShowFriendStatuses) return;
 			
 			try {
@@ -160,10 +178,13 @@ namespace DoTwitter
 					}
 				}
 			} catch { }
+			System.Console.WriteLine("UpdateTweets, end");
 		}
 				
 		private static void DownloadBuddyIcon (Uri imageUri, string location)
 		{
+			System.Console.WriteLine("DownloadBuddyIcon");
+
 			WebClient client = new WebClient ();
 			if (!System.IO.Directory.Exists (photo_directory))
 				System.IO.Directory.CreateDirectory (photo_directory);
@@ -172,6 +193,7 @@ namespace DoTwitter
 			} catch {
 				Console.Error.WriteLine ("{0} does not exist!",location);
 			}
+			System.Console.WriteLine("DownloadBuddyIcon end");
 		}
 			
 	}
