--- configure.ac.orig	2008-02-22 16:39:35.000000000 +0100
+++ configure.ac	2008-03-13 22:21:18.000000000 +0100
@@ -18,14 +18,6 @@
 AC_PROG_LIBTOOL
 
 dnl ==============================================
-dnl gconf stuff
-dnl ==============================================
-AC_PATH_PROG(GCONFTOOL, gconftool-2, no)
-if test x"$GCONFTOOL" = xno; then
-  AC_MSG_ERROR([gconftool-2 executable not found in your path - should be installed with GConf])
-fi
-
-dnl ==============================================
 dnl translation support
 dnl ==============================================
 
@@ -243,6 +235,13 @@
 dnl ==============================================
 dnl GConf
 dnl ==============================================
+if test "$with_gconf" != "no"; then
+AC_PATH_PROG(GCONFTOOL, gconftool-2, no)
+  if test x"$GCONFTOOL" = xno; then
+    AC_MSG_ERROR([gconftool-2 executable not found in your path - should be installed with GConf])
+  fi
+fi
+
 AM_GCONF_SOURCE_2
 
 dnl Please keep in alphabetical order
